hnn_core/opt_toy_example/general.py

```diff
@@ -0,0 +1,378 @@
+    return init_params, cons
+
+
+# gets called only once
+def _get_params_cobyla(net, constraints):
+    """Assembles constraints & initial parameters as required by fmin_cobyla.
+
+        Returns
+        -------
+        init_params : list
+        cons : list
+    """
+
+    init_params = list()
+    cons = list()
+
+    for drive_name in net.external_drives.keys():
+        for cons_idx, cons_key in enumerate(constraints[drive_name]):
+            if cons_key == 'mu' or cons_key == 'sigma':
+                cons.append(lambda x, idx=cons_idx:
+                            constraints[drive_name][cons_key][1] - x[idx])
+                cons.append(lambda x, idx=cons_idx:
+                            constraints[drive_name][cons_key][0] - x[idx])
+                init_params.append(
+                    net.external_drives[drive_name]['dynamics'][cons_key])
+            elif cons_key == 'weights':
+                conn_idxs = pick_connection(net, src_gids=drive_name)
+                for conn_idx in conn_idxs:
+                    cons.append(lambda x, idx=cons_idx:
+                                constraints[drive_name][cons_key][1] - x[idx])
+                    cons.append(lambda x, idx=cons_idx:
+                                constraints[drive_name][cons_key][0] - x[idx])
+                    init_params.append(
+                        net.connectivity[conn_idx]['nc_dict']['A_weight'])
+    return init_params, cons
+
+
+def _run_opt_bayesian(net, init_params, cons, metric, target_statistic,
+                      window_len):
+
+    @use_named_args(dimensions=cons)
+    def _obj_func(**params):
+        return metric(net, target_statistic, window_len, **params)
+
+    max_iter = 11
+    opt_results = gp_minimize(func=_obj_func,
+                              dimensions=cons,
+                              acq_func='EI',
+                              n_calls=max_iter,
+                              x0=init_params,
+                              random_state=64)
+    opt_params = opt_results.x
+    # get net_
+    # ...
+    error_values = [np.min(opt_results.func_vals[:i])
+                    for i in range(1, max_iter + 1)]
+    return opt_params, error_values
+
+
+def _run_opt_cobyla(net, init_params, cons, metric, target_statistic,
+                    window_len):
+
+    def _obj_func(params_cobyla):
+        return metric(net=net, target_statistic=target_statistic,
+                      window_len=window_len, params_cobyla=params_cobyla)
+
+    max_iter = 11
+    opt_results = fmin_cobyla(_obj_func,
+                              cons=cons,
+                              rhobeg=0.1,
+                              rhoend=1e-4,
+                              x0=init_params,
+                              maxfun=max_iter,
+                              catol=0.0)
+    opt_params = opt_results
+    # get net_
+    # ...
+    error_values = list()
+    return opt_params, error_values
+
+
+def _get_fixed_params(net):
```

Comment:
I don't quite get the logic of extracting the fixed params. Why can't we just pass around the network object and never worry about the fixed params?

Reply:
This is getting at the same issue I was pointing out in https://github.com/jonescompneurolab/hnn-core/pull/652#discussion_r1232398759.